File: bindings/python/examples/client/08_data_block.py

import json
import os
from dataclasses import asdict
from dotenv import load_dotenv
from iota_sdk import BasicBlockBody, Bip44, CoinType, Client, utf8_to_hex, hex_to_utf8, TaggedDataPayload, MnemonicSecretManager, SecretManager

load_dotenv()

node_url = os.environ.get('NODE_URL', 'https://api.testnet.shimmer.network')
issuer_id = os.environ.get(
    'ISSUER_ID', '0x0000000000000000000000000000000000000000000000000000000000000000')

if 'MNEMONIC' not in os.environ:
    raise Exception(".env MNEMONIC is undefined, see .env.example")

secret_manager = SecretManager(MnemonicSecretManager(os.environ['MNEMONIC']))

# Create a Client instance
client = Client(nodes=[node_url])

chain = Bip44(CoinType.IOTA)

# Create and post a block with a tagged data payload
unsigned_block = client.build_basic_block(
    issuer_id,
    TaggedDataPayload(
        utf8_to_hex("tag"),
        utf8_to_hex("data")))[0]
block = secret_manager.sign_block(unsigned_block, chain)
block_id = client.post_block(block)

print(f'Data block sent: {os.environ["EXPLORER_URL"]}/block/{block_id}')

block = client.get_block(block_id).block

if isinstance(block, BasicBlockBody):
    print(f'Block data: {json.dumps(asdict(block), indent=4)}')

    payload = block.payload

    if payload and 'data' in payload and payload['data']:
        print(f'Decoded data: { hex_to_utf8(payload["data"]) }')
else:
    raise ValueError("block must be an instance of BasicBlock")

File: bindings/nodejs/examples/client/08-data-block.ts
// Copyright 2021-2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {
    CoinType,
    Client,
    hexToUtf8,
    initLogger,
    TaggedDataPayload,
    utf8ToHex,
    SecretManager,
} from '@iota/sdk';
require('dotenv').config({ path: '.env' });

// Run with command:
// yarn run-example ./client/08-data-block.ts

// In this example we will send a block with a tagged data payload.
async function run() {
    initLogger();
    for (const envVar of ['NODE_URL', 'EXPLORER_URL']) {
        if (!(envVar in process.env)) {
            throw new Error(`.env ${envVar} is undefined, see .env.example`);
        }
    }

    if (!process.env.MNEMONIC) {
        throw new Error('.env MNEMONIC is undefined, see .env.example');
    }
    const mnemonicSecretManager = {
        mnemonic: process.env.MNEMONIC,
    };

    const secretManager = SecretManager.create(mnemonicSecretManager);

    const client = await Client.create({
        // Insert your node URL in the .env.
        nodes: [process.env.NODE_URL as string],
    });

    const issuerId = process.env.ISSUER_ID
        ? process.env.ISSUER_ID
        : '0x0000000000000000000000000000000000000000000000000000000000000000';

    const chain = {
        coinType: CoinType.IOTA,
        account: 0,
        change: 0,
        addressIndex: 0,
    };

    try {
        // Create block with tagged payload
        const unsignedBlock = await client.buildBasicBlock(
            issuerId,
            new TaggedDataPayload(utf8ToHex('Hello'), utf8ToHex('Tangle')),
        );
        const block = await secretManager.signBlock(unsignedBlock, chain);
        const blockId = await client.postBlock(block);

        console.log(`Block sent: ${process.env.EXPLORER_URL}/block/${blockId}`);

        const fetchedBlock = await client.getBlock(blockId);
        console.log('Block data: ', fetchedBlock);

        if (fetchedBlock.isBasic()) {
            const basic = fetchedBlock.asBasic();
            if (basic.payload instanceof TaggedDataPayload) {
                const payload = basic.payload as TaggedDataPayload;
                console.log('Decoded data:', hexToUtf8(payload.data));
            }
        }
    } catch (error) {
        console.error('Error: ', error);
    }
}

void run().then(() => process.exit());
